webui/react/src/stores/workspaces.tsx

```diff
@@ -1,69 +1,72 @@
 
-type WorkspacesContext = {
-  updateWorkspaces: (fn: (ws: Loadable<Workspace[]>) => Loadable<Workspace[]>) => void;
-  workspaces: Loadable<Workspace[]>;
-};
+type WorkspacesContext = { workspaces: WritableObservable<Loadable<Workspace[]>> };
 
 const WorkspacesContext = createContext<WorkspacesContext | null>(null);
 
 export const WorkspacesProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [workspaces, updateWorkspaces] = useState<Loadable<Workspace[]>>(NotLoaded);
-  return (
-    <WorkspacesContext.Provider value={{ updateWorkspaces, workspaces }}>
-      {children}
-    </WorkspacesContext.Provider>
-  );
+  const workspaces: WritableObservable<Loadable<Workspace[]>> = observable(NotLoaded);
+
+  return <WorkspacesContext.Provider value={{ workspaces }}>{children}</WorkspacesContext.Provider>;
 };
 
 export const useFetchWorkspaces = (canceler: AbortController): (() => Promise<void>) => {
   const context = useContext(WorkspacesContext);
+
   if (context === null) {
     throw new Error('Attempted to use useFetchWorkspaces outside of Workspace Context');
   }
-  const { updateWorkspaces } = context;
 
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getWorkspaces({}, { signal: canceler.signal });
-      updateWorkspaces(() => Loaded(response.workspaces));
-    } catch (e) {
-      handleError(e);
-    }
-  }, [canceler, updateWorkspaces]);
-};
-
-export const useEnsureWorkspacesFetched = (canceler: AbortController): (() => Promise<void>) => {
-  const context = useContext(WorkspacesContext);
-  if (context === null) {
-    throw new Error('Attempted to use useEnsureWorkspacesFetched outside of Workspace Context');
-  }
-  const { workspaces, updateWorkspaces } = context;
+  const { workspaces } = context;
 
   return useCallback(async (): Promise<void> => {
-    if (workspaces !== NotLoaded) return;
     try {
       const response = await getWorkspaces({}, { signal: canceler.signal });
-      updateWorkspaces(() => Loaded(response.workspaces));
+
+      workspaces.set(Loaded(response.workspaces));
     } catch (e) {
       handleError(e);
     }
-  }, [canceler, workspaces, updateWorkspaces]);
+  }, [canceler, workspaces]);
 };
 
 export const useWorkspaces = (params?: GetWorkspacesParams): Loadable<Workspace[]> => {
   const context = useContext(WorkspacesContext);
+
   if (context === null) {
     throw new Error('Attempted to use useWorkspaces outside of Workspace Context');
   }
-  return Loadable.map(context.workspaces, (workspaces: Workspace[]) =>
+
+  const { workspaces } = context;
+  const [workspacesState, setWorkspacesState] = useState(workspaces.get());
+
+  useEffect(() => {
+    const unsubscribe = workspaces.subscribe((ws, prevWs) => {
+      if (!isEqual(ws, prevWs)) setWorkspacesState(ws);
+    });
+
+    return () => {
+      unsubscribe();
+    };
+  }, [workspaces]);
+
+  return Loadable.map(workspacesState, (workspaces: Workspace[]) =>
```

Comment:
suggestion, non-blocking: if you do the `isEqual` check before `workspaces.set(Loaded(response.workspaces))`, then you can do `this.workspacesPublic = this.workspaces.select(ws => Loadable.map(ws, ...))`, (not a great name lol) in the provider itself, instead of doing all this again for each place `useWorkspaces` is called. bit of a nit, take it or leave it.